### ISSUE TITLE
Limiting demand in subquery may lock too many rows

### ISSUE THREAD

**Issue by mbuhot** [sorentwo/oban#4]
Oban may be susceptible to locking too many rows if the postgres query planner decides not to materialize the subquery used [here](https://github.com/sorentwo/oban/blob/master/lib/oban/query.ex#L32-L40).

There is a good write up of the problem [here](https://github.com/feikesteenbergen/demos/blob/master/bugs/update_from_correlated.adoc)

I haven't reproduced this problem in Oban, just browsed the code and noticed the structure looked similar to that from the linked article.

**Comment by sorentwo** [comment]
Thanks for sharing the link and your concern, that is extremely good to know. So far in my usage and testing I haven't noticed the problem. I wonder if the additional complexity of the query prevents the failure conditions or if the orphaned job rescue hides it?

I'll do some experiments. Switching to use a CTE seems like a really clean solution but there isn't any DSL support in Ecto yet.

**Comment by sorentwo** [comment]
I took a little time to check into this and it doesn't _seem_ to be a problem.

Here is a slightly modified variant of the query with the arguments replaced by literal values and a shorter return statement:

```sql
UPDATE "oban_jobs" AS o0
SET "state" = 'executing', "attempted_at" = now(), "attempt" = o0."attempt" + 1
FROM (
  SELECT o0."id" AS "id", pg_try_advisory_lock_shared(o0."id") AS "lock"
  FROM "oban_jobs" AS o0
  WHERE (o0."state" = 'available')
  AND (o0."queue" = 'default')
  AND (o0."scheduled_at" <= now())
  ORDER BY o0."scheduled_at", o0."id"
  LIMIT 2
  FOR UPDATE SKIP LOCKED
) AS s1
WHERE (o0."id" = s1."id")
RETURNING o0."id";
```

With seven jobs put into the `default` queue this is what I see for the query plan:

```
                                                                                   QUERY PLAN
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
 Update on oban_jobs o0  (cost=8.07..16.25 rows=1 width=638) (actual time=0.126..0.162 rows=2 loops=1)
   ->  Nested Loop  (cost=8.07..16.25 rows=1 width=638) (actual time=0.101..0.125 rows=2 loops=1)
         ->  Subquery Scan on s1  (cost=7.92..7.96 rows=1 width=40) (actual time=0.080..0.099 rows=2 loops=1)
               ->  Limit  (cost=7.92..7.95 rows=1 width=23) (actual time=0.076..0.093 rows=2 loops=1)
                     ->  LockRows  (cost=7.92..7.95 rows=1 width=23) (actual time=0.075..0.091 rows=2 loops=1)
                           ->  Result  (cost=7.92..7.94 rows=1 width=23) (actual time=0.052..0.066 rows=2 loops=1)
                                 ->  Sort  (cost=7.92..7.93 rows=1 width=22) (actual time=0.033..0.034 rows=2 loops=1)
                                       Sort Key: o0_1.scheduled_at, o0_1.id
                                       Sort Method: quicksort  Memory: 25kB
                                       ->  Index Scan using oban_jobs_queue_index on oban_jobs o0_1  (cost=0.14..7.91 rows=1 width=22) (actual time=0.009..0.022 rows=7 loops=1)
                                             Index Cond: (queue = 'default'::text)
                                             Filter: (state = 'available'::oban_job_state)
                                             Rows Removed by Filter: 2
         ->  Index Scan using oban_jobs_pkey on oban_jobs o0  (cost=0.14..8.16 rows=1 width=594) (actual time=0.009..0.009 rows=1 loops=2)
               Index Cond: (id = s1.id)
 Planning Time: 0.330 ms
```

The actual row counts look correct. After three invocations of the query the counts grouped by state look correct:

```
[local]/oban_test=# select state, count(*) from oban_jobs group by state;
   state   │ count
═══════════╪═══════
 executing │     6
 available │     3
```

Thankfully there doesn't seem to be a problem! Thanks for letting me know about it though, it was a nice learning experience.

**Comment by speeddragon** [comment]
Not sure if related, but since I've started used `oban` I've been getting a lot of this error.

```
SQLSTATE[53200]: Out of memory: 7 ERROR: out of shared memory HINT: You might need to increase max_locks_per_transaction.
```

I've updated the `max_locks_per_transaction` from 64 to 128, but not sure if this is normal.

**Comment by sorentwo** [comment]
@speeddragon That _shouldn't_ be related, but I'd like to understand what is going on for you just to be sure. I've done some more research on this and experimented more with the query and arbitrary shared transactional locks. So far I can't recreate the issue.

A few questions:

1. What is your db pool size?
2. How many tables are in the database?
3. Do you have any background jobs touching a large number of tables within a transaction?

**Comment by speeddragon** [comment]
1. 15 on the elixir app. I've a PHP app that it is using around 20.
2. 189 tables.
3. So, before creating this elixir app, I've some scripts (~20) running in background with PHP. I never had this issue before. I'm starting migrating this jobs to Elixir and I've decided to use Oban. I've worked in other Elixir projects, and never had this issue before.

Hope this helps.

**Comment by sorentwo** [comment]
@speeddragon Thanks for the quick info. I'm glad you were able to alleviate the issue by increasing the value. Unfortunately the `max_locks_per_transaction` setting can't be set after boot, and environments like Heroku don't allow you to modify the value.

My max locks was set to 1024 for some reason that I don't remember anymore. By reducing it I've managed to reproduce the issue locally and am working on a fix now.

**Comment by sorentwo** [comment]
@speeddragon Thanks again for reporting this. There is a proper fix for the issue in ad5ccd5 💛
